### PR TITLE
Modernized disease contraction code and made biosuits more useful.

### PR DESCRIPTION
**Pull request by jaw-sh**
## About The Pull Request
This PR answers a question of how to make bio suits more useful by modernizing disease contraction code. Instead of manually checking specific articles of clothing and their permeability, the logic now uses the mob's equipment's `body_parts_covered` flags. This code works for all mobs and does not require new snowflake code for different kinds of mobs which may have different equipment slots (i.e. monkeys and humans now use the same code).

## Why It's Good For The Game
Viruses are hard to deal with for medical staff because curing them often means getting yourself infected first. Bio suits, surgical masks, and latex gloves have often simply been cosmetics. I have verified all these items now have very real protective qualities in stopping the spread of disease. Bio suits especially completely prevent the contraction of airborne disease via the air.

## Changelog
:cl:
refactor: Modernized disease contraction logic.
balance: Bio suits now offer absolute protection from diseases. Bio hoods especially stop airborne contraction.
add: Added new medical tool "bag valve mask" which expediently performs CPR without virus contamination risks, even without lungs or the ability to breathe, or while wearing headgear.
fix: Fixed _adjusted_ masks providing 100% protection against viruses.
/:cl:

### CODE REVIEW

**Comment by jaw-sh** [comment]
My last commit extends this a bit further.

CPR now spreads viruses, up to and including fluid contact. If either the giver or receiver has an injured head, then blood exposure is also done. To safely perform CPR with an infected patient, you can now use an "air valve bag". This was accomplished by moving some procs around.
image

The air valve bag is not just a solution to a problem that I created. It also provides a way to perform CPR on a patient if you have a virus that makes it so you do not breathe, or otherwise do not have functional lungs.

**Comment by Tlaltecuhtli** [comment]
does wearing a surgical mask stop the virus from spreading?

**Comment by jaw-sh** [comment]
> does wearing a surgical mask stop the virus from spreading?
Surgical masks provide excellent protection from airborne viruses when worn properly. It is technically 99%, but airborne contraction checks proc very frequently around sick mobs so it's not effective immunity.

**Comment by MimicFaux** [comment]
I support this PR for the bag valve mask alone. Nice to see some robustness being added into the viral systems.

**Comment by msgerbs** [comment]
Should the bag valve mask be somewhat less effective than proper CPR as a tradeoff for its benefits?

File: code/modules/surgery/tools.dm [hunk 1]
```diff
@@ -442,3 +442,33 @@
...
+	if(!iscarbon(M) || !iscarbon(user) || user.a_intent != INTENT_HELP)
+		return ..()
+
+	var/mob/living/carbon/CM = M
+	var/mob/living/carbon/CU = user
+
+	if(check_zone(CU.zone_selected) != BODY_ZONE_HEAD)
+		return ..()
+
+	if(CM.body_position != LYING_DOWN  || (CM.health >= 0 && !HAS_TRAIT(CM, TRAIT_FAKEDEATH)))
```
Comment (ATH1909) [inline code/modules/surgery/tools.dm:471]: CPR checks for crit_threshold, but this checks for your health being >= 0...?

Comment (jaw-sh) [reply code/modules/surgery/tools.dm:471]: Whoops, that's copied from `/datum/species/proc/help` and is meant to be inverted. I'll fix that logic and update.

**Comment by LemonInTheDark** [comment]
I smells a feature, please don't add stuff in your fix pr, it makes this so much more painful to deal with.

**Comment by jaw-sh** [comment]
> Please don't add new items like the bag valve in your fix pr, we're in a feature freeze right now, and that conflicts with it.

A significant part of that commit is changes to procs I'd like in upstream due to potential merge conflicts down the road. Is removing the item sufficient?

**Comment by ExcessiveUseOfCobblestone** [comment]
Your PR is being blocked/closed due to an ongoing freeze. Please review the appropriate PR on the tracker to learn more as to what is/is not appropriate for the freeze.

**Comment by jaw-sh** [comment]
> Your PR is being blocked/closed due to an ongoing freeze. Please review the appropriate PR on the tracker to learn more as to what is/is not appropriate for the freeze.

As I have explained before,
I need to move procs. I would like these proc moves upstream. This will help with merging. I would like to submit the proc moves with the rest of these changes. If I remove just the item, is that sufficient?

**Comment by Krysonism** [comment]
The first good thing you've done since the Bulla video.

File: code/modules/surgery/tools.dm (hunk 1 above)
Comment (jaw-sh) [reply code/modules/surgery/tools.dm]: Per 7ba1f6b, both CPR checks now prefer the `crit_threshold` property over static 0.

**Comment by LemonInTheDark** [comment]
> I need to move procs. I would like these proc moves upstream. This will help with merging. I would like to submit the proc moves with the rest of these changes. **If I remove just the item, is that sufficient?**

Which procs? viral_contact? or the support for `tool_assisted`

**Comment by jaw-sh** [comment]
> > I need to move procs. I would like these proc moves upstream. This will help with merging. I would like to submit the proc moves with the rest of these changes. **If I remove just the item, is that sufficient?**
> 
> Which procs? viral_contact? or the support for `tool_assisted`

The support for tool assisted is the big one. Anything that changes files especially because Git can't handle that well.

**Comment by ExcessiveUseOfCobblestone** [comment]
If it's not a strict fix or code improvement to current functionality it needs to go (for now). For instance if you should not be changing the overall accessibility to contracting diseases unless it's because the current implementation is buggy and you simply cannot fix it without consequently touching the balance.

Other items include the strength of items in terms of avoiding diseases, and new features like the cpr bag.

**Comment by jaw-sh** [comment]
I can't promise that even if I undo the bag and the biosuit buff that the contractions will work the same for practical purposes. In my experience, viruses have no trouble spreading at all, but there may be certain combinations of clothing that offer stronger multiplicative protection now due to coverage checks than before. That's entirely possible and I'm sure there's at least one situation where that is the case.

I'd obviously still prefer these changes be upstream because they would be difficult to maintain and handle merges with but if you want to wait until December I would be fine with that.

**Comment by LemonInTheDark** [comment]
That might be better, I'd rather not have dead code waiting for a promised feature downline, seems like a poor pattern to follow. It also removes the shadow of feature, so this change can be properly discussed.

**Comment by ExcessiveUseOfCobblestone** [comment]
just chill until the freeze is over for this then

**Comment by jaw-sh** [comment]
Happy first of December @ExcessiveUseOfCobblestone. What should I do if I want to resubmit this?

**Comment by InvalidArgument3** [comment]
Hope this PR gets reopened, it seemed very neat.

**Comment by ExcessiveUseOfCobblestone** [comment]
@jaw-sh re pr it

**Pull request closed without merging**